nakamoto-p2p/src/prototype.rs

```diff
@@ -0,0 +1,888 @@
+pub mod protocol {
+    use log::*;
+
+    use crate::error::Error;
+    use crate::peer::{Config, Link};
+    use crate::PeerId;
+
+    use std::collections::{HashMap, HashSet};
+    use std::fmt::Debug;
+    use std::net;
+    use std::time::{self, SystemTime, UNIX_EPOCH};
+
+    use bitcoin::network::address::Address;
+    use bitcoin::network::constants::ServiceFlags;
+    use bitcoin::network::message::{NetworkMessage, RawNetworkMessage};
+    use bitcoin::network::message_blockdata::GetHeadersMessage;
+    use bitcoin::network::message_network::VersionMessage;
+    use bitcoin::util::hash::BitcoinHash;
+
+    use nakamoto_chain::block::time::AdjustedTime;
+    use nakamoto_chain::block::tree::BlockTree;
+    use nakamoto_chain::block::{BlockHash, BlockHeader, Height};
+
+    /// User agent included in `version` messages.
+    pub const USER_AGENT: &str = "/nakamoto:0.0.0/";
+    /// Duration of inactivity before timing out a peer.
+    pub const IDLE_TIMEOUT: time::Duration = time::Duration::from_secs(60 * 5);
+    /// How long to wait between sending pings.
+    pub const PING_INTERVAL: time::Duration = time::Duration::from_secs(60);
+    /// Number of blocks out of sync we have to be to trigger an initial sync.
+    pub const SYNC_THRESHOLD: Height = 144;
+    /// Minimum number of peers to be connected to.
+    pub const PEER_CONNECTION_THRESHOLD: usize = 3;
+    /// Maximum time adjustment between network and local time (70 minutes).
+    pub const MAX_TIME_ADJUSTMENT: TimeOffset = 70 * 60;
+
+    /// A time offset, in seconds.
+    pub type TimeOffset = i64;
+
+    /// A finite-state machine that can advance one step at a time, given an input event.
+    /// Parametrized over the message type.
+    pub trait Protocol<M> {
+        /// Process the next event and advance the state-machine by one step.
+        /// Returns messages destined for peers.
+        fn step(&mut self, event: Event<M>) -> Vec<(PeerId, M)>;
```

Comment:
An improvement which might address the TODO further down about leakage of io and socket into the protocol could be to rework this to take a `State` - which could be parameterised as well - and outputs the new state and the list of messages. Then you can asssert the correctness of the protocol state machine in a way that doesn't need to manage side-effects like sockets and their io. And the responsibility of the reactor (or however the component is called) is to hold onto the state, feed the input events and disseminate the outputs, including updating it's stored state.